### ISSUE TITLE
collection.findOne() called twice

### ISSUE THREAD

**Issue by gentunian** [meteor/meteor#4508]
I'm wondering why `collection.findOne()` is being called twice. This behaviour is not happening with `collection.find()` method. I've created a gist here:

https://gist.github.com/gentunian/c0a3d8a755c86636a93d

It's also interesting what happens when you change the order of the helpers:

    <template name="testTemplate">
    <p>{{testFindOneHelper}}</p>
    <p>{{testFindHelper}}</p>
    </template> 

outputs this:

    [testFindOneHelper]Wed Jun 03 2015 20:07:08 GMT-0300 (ART)
    [testFindOneHelper]undefined
    [testFindHelper]Wed Jun 03 2015 20:07:08 GMT-0300 (ART)
    [testFindHelper][object Object]
    [testFindOneHelper]Wed Jun 03 2015 20:07:09 GMT-0300 (ART)
    [testFindOneHelper][object Object]

and this: 

    <template name="testTemplate">
    <p>{{testFindHelper}}</p>
    <p>{{testFindOneHelper}}</p>
    </template>

outputs this:

    [testFindHelper]Wed Jun 03 2015 20:06:18 GMT-0300 (ART)
    [testFindHelper][object Object]
    [testFindOneHelper]Wed Jun 03 2015 20:06:18 GMT-0300 (ART)
    [testFindOneHelper]undefined
    [testFindOneHelper]Wed Jun 03 2015 20:06:18 GMT-0300 (ART)
    [testFindOneHelper][object Object]

The output is rather different. In the first example, the helper order is not what I would expect.

**Comment by robertpitt** [comment]
I think I see what's going on here, correct me if I am wrong.

Your Mongo collection has two records in, each record is sent over the DDP connection is individual packets, each time the client receives a packet the collection event handler triggers the re-computation 
 of your template.

The first time your template is rendered, the data has not yet reached the client, hence the undefined, once the first data has reached the client your template is re-computed, then the second packet is received causing the second re-computation.

your `testFindHelper` is calling `find({})`, which means that the tracker monitors documents coming down the pipe that match `{}`, hence the two re-computations.

and so your findOne will only trigger on the first packet.

TL/DR

Your `findOne({})` will cause one re-computation
your `find({})` will cause **n** re-computations, where n is the number of documents exposed to the client in the users collection.

**Comment by gentunian** [comment]
@robertpitt Almost like that, based on an IRC answer I got from #meteor channel:

> **oRiCLe:** gentunian: once when the template is rendered, and once when the subscription fills the collection on the client...
> **oRiCLe:** if you want to prevent, make sure the subscription is in the ready state before displaying the template.
> **gentunian:** oRiCLe, I found that but why that doesn't happens with find() ?
> **oRiCLe:** find creates an observe cursor

Despite your explanation seems ok, `findOne()` is not called N times. It was just a mere coincidence that I had two records, but I have the same results with more than 2 records.

**Comment by ekatek** [comment]
Looks like IRC figured this one out! Thanks all and good luck with your Meteor development.
